sys/drv/rtl8139.c

```diff
@@ -32,18 +50,133 @@ static int rtl_reset(rtl8139_state_t *state) {
+  uint16_t status = bus_read_2(state->regs, RL_ISR);
+
+  if (status & RL_ISR_RX_OK)
+    do_receive(state);
+  else if (status & RL_ISR_TX_OK)
+    do_transceive(state);
+
+  return IF_FILTERED;
+}
+
+static int set_mem_contig_address(rtl8139_state_t *state) {
+  paddr_t page2; /* address of second allocated page */
+
+  /* set and check if any address is assigned */
+  if (!pmap_kextract(state->rx_buf, &state->paddr))
+    return -1;
+
+  /* check if memory is contiguous */
+  if (!pmap_kextract(state->rx_buf + PAGESIZE, &page2))
+    return -1;
+
+  if (state->paddr + PAGESIZE != page2)
+    return -1;
+
+  return 0;
+}
+
+static void skb_free(struct sk_buff *skb) {
+  if (skb->length)
+    kfree(M_DEV, skb->data);
+  kfree(M_DEV, skb);
+}
+
+static int rtl8139_read(vnode_t *v, uio_t *uio, int ioflag) {
+  rtl8139_state_t *state = devfs_node_data(v);
+  /* temporary workaround to stop fread */
+  static int fake_0 = 0;
+  /* TODO: WITH_MTX_LOCK { */
+  uio->uio_offset = 0;
+
+  if (fake_0) {
+    fake_0 = 0;
+    return uiomove_frombuf(NULL, 0, uio);
+  }
+
+  struct sk_buff *skb = TAILQ_FIRST(&state->rx_queue);
+  if (skb) {
+    int res = uiomove_frombuf(skb->data, skb->length, uio);
+    TAILQ_REMOVE(&state->rx_queue, skb, queue);
+    skb_free(skb);
+    fake_0 = 1;
+    return res;
+  } else
+    return uiomove_frombuf(NULL, 0, uio);
+}
+
+static vnodeops_t rtl8139_vnodeops = {.v_read = rtl8139_read};
+
+int rtl8139_attach(device_t *dev) {
   rtl8139_state_t *state = dev->state;
 
+  pci_enable_busmaster(dev);
   state->regs = device_take_memory(dev, 1, RF_ACTIVE);
-  if (state->regs == NULL)
+
+  /*
+   * XXX: there is no irq routing in mimiker
+   *      so in this case we need to add irq manually
+   */
+  device_add_irq(dev, 0, 10);
+  state->irq_res = device_take_irq(dev, 0, RF_ACTIVE);
+  if (!state->regs || !state->irq_res) {
+    klog("rtl8139 failed to init resources");
+    return ENXIO;
+  }
+  bus_intr_setup(dev, state->irq_res, rtl8139_intr, NULL, state, "RTL8139");
+
+  TAILQ_INIT(&state->rx_queue);
+  TAILQ_INIT(&state->tx_queue);
+  state->rx_offset = 0;
+  state->rx_buf = (vaddr_t)kmem_alloc(2 * PAGESIZE, M_ZERO);
+  // TODO: mark this memory as PMAP_NOCACHE
+  if (set_mem_contig_address(state)) {
+    klog("failed to alloc contig memory");
     return ENXIO;
+  }
 
   if (rtl_reset(state)) {
     klog("RTL8139: Failed to reset device!");
     return ENXIO;
   }
 
+  /* set-up address for dma */
+  bus_write_4(state->regs, RL_RXADDR, state->paddr);
+  /* we want to receive all packets - promiscious mode */
+  /* 0xf - accept broadcast, multicast, physical match, all other packets
+   * (promisc) */
+  /* 0x80 - wrap bit */
+  bus_write_4(state->regs, RL_RXCFG, 0x8f);
+  bus_write_1(state->regs, RL_COMMAND, RL_CMD_RX_ENB | RL_CMD_TX_ENB);
+  // IMR - mask rx enable
+  bus_write_2(state->regs, RL_IMR, RL_ISR_RX_OK);
+
+  devfs_makedev(NULL, "rtl8139", &rtl8139_vnodeops, state, NULL);
   return 0;
```

Comment:
Please note in the description that some changes are dependent on another PR [890](https://github.com/cahirwpz/mimiker/pull/890).